jmespath/functions.go

```diff
@@ -0,0 +1,818 @@
+	if len(items) == 0 {
+		return nil, nil
+	}
+	if len(items) == 1 {
+		return items[0], nil
+	}
+	best := items[0]
+	for _, item := range items[1:] {
+		if item < best {
+			best = item
+		}
+	}
+	return best, nil
+}
+
+func jpfMinBy(arguments []interface{}) (interface{}, error) {
+	intr := arguments[0].(*treeInterpreter)
+	arr := arguments[1].([]interface{})
+	exp := arguments[2].(expRef)
+	node := exp.ref
+	if len(arr) == 0 {
+		return nil, nil
+	} else if len(arr) == 1 {
+		return arr[0], nil
+	}
+	start, err := intr.Execute(node, arr[0])
+	if err != nil {
+		return nil, err
+	}
+	if t, ok := start.(float64); ok {
+		bestVal := t
+		bestItem := arr[0]
+		for _, item := range arr[1:] {
+			result, err := intr.Execute(node, item)
+			if err != nil {
+				return nil, err
+			}
+			current, ok := result.(float64)
+			if !ok {
+				return nil, errors.New("Invalid type, must be number")
+			}
+			if current < bestVal {
+				bestVal = current
+				bestItem = item
+			}
+		}
+		return bestItem, nil
+	} else if t, ok := start.(string); ok {
+		bestVal := t
+		bestItem := arr[0]
+		for _, item := range arr[1:] {
+			result, err := intr.Execute(node, item)
+			if err != nil {
+				return nil, err
+			}
+			current, ok := result.(string)
+			if !ok {
+				return nil, errors.New("Invalid type, must be string")
+			}
+			if current < bestVal {
+				bestVal = current
+				bestItem = item
+			}
+		}
+		return bestItem, nil
+	} else {
+		return nil, errors.New("Invalid type, must be number of string.")
+	}
+}
+func jpfType(arguments []interface{}) (interface{}, error) {
+	arg := arguments[0]
+	if _, ok := arg.(float64); ok {
+		return "number", nil
+	}
+	if _, ok := arg.(string); ok {
+		return "string", nil
+	}
+	if _, ok := arg.([]interface{}); ok {
+		return "array", nil
+	}
+	if _, ok := arg.(map[string]interface{}); ok {
+		return "object", nil
+	}
+	if arg == nil {
+		return "null", nil
+	}
+	if arg == true || arg == false {
+		return "boolean", nil
+	}
+	return nil, errors.New("Unknown type")
+}
+func jpfKeys(arguments []interface{}) (interface{}, error) {
+	arg := arguments[0].(map[string]interface{})
+	collected := []interface{}{}
+	for key := range arg {
+		collected = append(collected, key)
+	}
+	return collected, nil
+}
+func jpfValues(arguments []interface{}) (interface{}, error) {
+	arg := arguments[0].(map[string]interface{})
+	collected := []interface{}{}
+	for _, value := range arg {
+		collected = append(collected, value)
+	}
+	return collected, nil
+}
+func jpfSort(arguments []interface{}) (interface{}, error) {
+	if items, ok := jputil.ToArrayNum(arguments[0]); ok {
+		d := sort.Float64Slice(items)
+		sort.Stable(d)
+		final := make([]interface{}, len(d))
+		for i, val := range d {
+			final[i] = val
+		}
+		return final, nil
+	}
+	// Otherwise we're dealing with sort()'ing strings.
+	items, _ := jputil.ToArrayStr(arguments[0])
+	d := sort.StringSlice(items)
+	sort.Stable(d)
+	final := make([]interface{}, len(d))
+	for i, val := range d {
```

Comment:
It compiles and appears to work.  However, this will cause failures in the compliance tests.  I'm not sure if there's a general way to handle this, but in the [function that runs the compliance tests](https://github.com/jmespath/jmespath.go/blob/initial-review/compliance_test.go#L71-L82) the JSON is parsed as an `interface{}`, so if I return  `[]string` from this function the types won't match you'll get this failure from `make test`:

```
--- FAIL: TestCompliance (0.03s)
        Error Trace:    compliance_test.go:120
			compliance_test.go:90
			compliance_test.go:78
			compliance_test.go:66
	Error:		Not equal: []interface {}{"bar", "foo"} (expected)
			        != []string{"bar", "foo"} (actual)
	Messages:	Expression: sort(keys(objects))

        Error Trace:    compliance_test.go:120
			compliance_test.go:90
			compliance_test.go:78
			compliance_test.go:66
	Error:		Not equal: []interface {}{"bar", "baz"} (expected)
			        != []string{"bar", "baz"} (actual)
	Messages:	Expression: sort(values(objects))

        Error Trace:    compliance_test.go:120
			compliance_test.go:90
			compliance_test.go:78
			compliance_test.go:66
	Error:		Not equal: []interface {}{"a", "b", "c"} (expected)
			        != []string{"a", "b", "c"} (actual)
	Messages:	Expression: sort(strings)
```

It makes sense why it fails, but I'm not sure what to do about it.  Is there a good way to handle this in a general way?